include/dfrn.php

```diff
@@ -0,0 +1,791 @@
+	$attributes = array(
+			"rel" => "photo",
+			"type" => "image/jpeg",
+			"media:width" => 80,
+			"media:height" => 80,
+			"href" => $contact["photo"]);
+	xml_add_element($doc, $author, "link", "", $attributes);
+
+	$attributes = array(
+			"rel" => "avatar",
+			"type" => "image/jpeg",
+			"media:width" => 80,
+			"media:height" => 80,
+			"href" => $contact["photo"]);
+	xml_add_element($doc, $author, "link", "", $attributes);
+
+	return $author;
+}
+
+/**
+ * @brief Adds the activity elements
+ *
+ * @param object $doc XML document
+ * @param string $element Element name for the activity
+ * @param string $activity activity value
+ *
+ * @return object XML activity object
+ */
+function dfrn_create_activity($doc, $element, $activity) {
+
+	if($activity) {
+		$entry = $doc->createElement($element);
+
+		$r = parse_xml_string($activity, false);
+		if(!$r)
+			return false;
+		if($r->type)
+			xml_add_element($doc, $entry, "activity:object-type", $r->type);
+		if($r->id)
+			xml_add_element($doc, $entry, "id", $r->id);
+		if($r->title)
+			xml_add_element($doc, $entry, "title", $r->title);
+		if($r->link) {
+			if(substr($r->link,0,1) === '<') {
+				if(strstr($r->link,'&') && (! strstr($r->link,'&amp;')))
+					$r->link = str_replace('&','&amp;', $r->link);
+
+				$r->link = preg_replace('/\<link(.*?)\"\>/','<link$1"/>',$r->link);
+
+				$data = parse_xml_string($r->link, false);
+				foreach ($data->attributes() AS $parameter => $value)
+					$attributes[$parameter] = $value;
+			} else
+				$attributes = array("rel" => "alternate", "type" => "text/html", "href" => $r->link);
+
+			xml_add_element($doc, $entry, "link", "", $attributes);
+		}
+		if($r->content)
+			xml_add_element($doc, $entry, "content", bbcode($r->content), array("type" => "html"));
+
+		return $entry;
+	}
+
+	return false;
+}
+
+/**
+ * @brief Adds the elements for attachments
+ *
+ * @param object $doc XML document
+ * @param object $root XML root
+ * @param array $item Item element
+ *
+ * @return object XML attachment object
+ */
+function dfrn_get_attachment($doc, $root, $item) {
+	$arr = explode('[/attach],',$item['attach']);
+	if(count($arr)) {
+		foreach($arr as $r) {
+			$matches = false;
+			$cnt = preg_match('|\[attach\]href=\"(.*?)\" length=\"(.*?)\" type=\"(.*?)\" title=\"(.*?)\"|',$r,$matches);
+			if($cnt) {
+				$attributes = array("rel" => "enclosure",
+						"href" => $matches[1],
+						"type" => $matches[3]);
+
+				if(intval($matches[2]))
+					$attributes["length"] = intval($matches[2]);
+
+				if(trim($matches[4]) != "")
+					$attributes["title"] = trim($matches[4]);
+
+				xml_add_element($doc, $root, "link", "", $attributes);
+			}
+		}
+	}
+}
+
+/**
+ * @brief Adds the "entry" elements for the DFRN protocol
+ *
+ * @param object $doc XML document
+ * @param string $type "text" or "html"
+ * @param array $item Item element
+ * @param array $owner Owner record
+ * @param bool $comment Trigger the sending of the "comment" element
+ * @param int $cid
```

Comment:
I guess so. It is used for the private photos - but it is part of the structure that I reused.